atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -457,34 +460,38 @@ public Iterator<Map.Entry<Cell, byte[]>> getRowsColumnRange(
 
         BatchColumnRangeSelection batchColumnRangeSelection =
                 BatchColumnRangeSelection.create(columnRangeSelection, batchHint);
-        return scopeToTransaction(getPostFilteredColumns(tableRef, batchColumnRangeSelection, rawResults));
+        return scopeToTransaction(getPostFilteredColumns(tableRef, batchColumnRangeSelection, stableRows, rawResults));
     }
 
     @Override
+    @SuppressWarnings("MustBeClosedChecker") // Sadly we can't close properly here without an ABI break :/
     public Map<byte[], Iterator<Map.Entry<Cell, byte[]>>> getRowsColumnRangeIterator(
             TableReference tableRef, Iterable<byte[]> rows, BatchColumnRangeSelection columnRangeSelection) {
         checkGetPreconditions(tableRef);
         if (Iterables.isEmpty(rows)) {
             return ImmutableMap.of();
         }
         hasReads = true;
-        Map<byte[], RowColumnRangeIterator> rawResults =
-                keyValueService.getRowsColumnRange(tableRef, rows, columnRangeSelection, getStartTimestamp());
-        ImmutableSortedMap.Builder<byte[], Iterator<Map.Entry<Cell, byte[]>>> postFilteredResultsBuilder =
-                ImmutableSortedMap.orderedBy(PtBytes.BYTES_COMPARATOR);
-        for (Map.Entry<byte[], RowColumnRangeIterator> e : rawResults.entrySet()) {
-            byte[] row = e.getKey();
-            RowColumnRangeIterator rawIterator = e.getValue();
-
-            // Sadly we can't close properly here without an ABI break :/
-            @SuppressWarnings("MustBeClosedChecker")
-            ClosableIterator<Map.Entry<Cell, byte[]>> postFilteredIterator =
-                    getPostFilteredColumns(tableRef, columnRangeSelection, row, rawIterator);
-
-            postFilteredResultsBuilder.put(row, scopeToTransaction(postFilteredIterator));
-        }
-        SortedMap<byte[], Iterator<Map.Entry<Cell, byte[]>>> postFilteredResults =
-                postFilteredResultsBuilder.buildOrThrow();
+        ImmutableSortedMap<byte[], RowColumnRangeIterator> rawResults = ImmutableSortedMap.copyOf(
```

Comment:
Just a note for future PR readers: this is _technically_ not needed, just because `getRowsColumnRange` happens to return all rows even if they do not exist, but this is _not_ explicitly stated in the API. This is the more defensive move which is better 👍🏼

Reply:
This isn't why we do this.

We do this because arrays implement `equals` using reference equality. Without this we would need the KVS to always return a map whose keys are the same instances as those in the `rows` parameter.

See https://github.com/palantir/atlasdb/pull/6424#discussion_r1085756186

Reply:
In hindsight + with a bit of testing I think I've determined all KVSes actually implement KVS getRowsColumnRange with an empty iterator already (though the API does not say anything), so it's a bit naughty but I'll just hard-define returning empty iterator as *required* for KVSes, which avoids the need for this change at all

Reply:
Ah, thanks for clarifying -- that's very cheeky! Makes this change all the more defensive.